docs/tutorial/creating_cli.md

```diff
@@ -0,0 +1,622 @@
+```
+But of course, you can use any format you want, like JSON, CSV, etc.
+
+Once we defined the encoding, we need a function to parse the file representation to our type in Scala:
+
+```scala 3
+def impl(bookPath: Path): ContactManager = new ContactManager:
+  
+    private def parseContact(contact: String): Either[Exception, Contact] =
+      contact.split('|') match 
+        case Array(id, firstName, lastName, phoneNumber, email) =>
+          Contact(id, firstName, lastName, phoneNumber, email).asRight
+        case _ => new Exception(s"Invalid contact format: $contact").asLeft
+    end parseContact
+```
+
+Next, a function to encode the contact to a string:
+
+```scala 3
+    private def encodeContact(contact: Contact): String =
+      s"${contact.username}|${contact.firstName}|${contact.lastName}|${contact.phoneNumber}|${contact.email}"
+```
+
+Now, we can implement the `addContact` function:
+
+```scala 3
+override def addContact(contact: Contact): IO[Username] = 
+  bookPath.readLines.flatMap: lines =>               // (1)
+    lines.find(_.startsWith(contact.username)) match // (2)
+      case Some(_) =>
+        ContactFound(contact.username).raiseError[IO, Username]
+      
+      case None =>
+        lines.isEmpty
+          .pure[IO]
+          .ifM( // (3)
+            bookPath.write(encodeContact(contact)).as(contact.username),  // (4)
+            bookPath.appendLine(encodeContact(contact)).as(contact.username)
+          )
+          
+end addContact
+```
+
+First, we read all of our lines, line by line using the `FilesOs#readLines` method (1). Then, we check if the contact already exists in the file checking if the username is contained (2). If it does, we raise an error using the `raiseError` method. If it doesn't, we check if the file is empty (3) using the `ifM`; this method basically checks if the boolean inside the `IO` is true or false and it executes an effect depending on the condition. If it is, we write the contact to the file using the `write` method. If it's not, we append the contact to the file using the `appendLine` method (4).
+
+
+Now that we have the `addContact` function implemented, we can move to the `removeContact` function:
+
+```scala 3
+override def removeContact(username: Username): IO[Unit] =
+  for
+    contacts <- bookPath.readLines
+    _ <- bookPath.writeLines(
+      contacts.filterNot(_.startsWith(username))
+    )
+  yield ()
+```
+
+The explanation is straightforward. We load all the contacts from the file, and then we filter the ones that don't match the username we want to remove. Finally, we write the filtered contacts back to the file.
+
+Now is the turn of the search functions. The first one is the `searchId` function:
+
+```scala 3
+override def searchId(username: Username): IO[Option[Contact]] =
+  for 
+    contacts <- bookPath.readLines
+  yield contacts
+    .find(_.startsWith(username))
+    .flatMap(parseContact(_).toOption)
```

Comment:
Why is it called `searchId` if it searches by `username`?

Reply:
In my mind, a username is unique in a database (just like an id). So, to avoid writing `searchByUsername` (quite lonmng) I  leaved it `searchId`. Should I change everywhere? :o

Reply:
It wouldn't hurt. Let me be paternalistic for a moment: using a username, an email or any user submitted data as an id in a database is a terrible idea. Ids in general are meant to be unique and never changing (unlike a username), and as such they should be generated by the application's backend.

Reply:
It's better to rename it as `searchUsername` and modify all the commands accordingly